### PR TITLE
41 tensorboard logging

### PR DESCRIPTION
**Pull request by nvkevlu**
Screenshot from 2020-02-05 10-55-09
Initial MVP to log to tensorboard. Right now there is nothing for validation so I am logging the loss as a scalar and animated gifs of the images. With validation, we can create a contour with cv2 and combine it with the image then save it the same way.

### CODE REVIEW

File: examples/unet_segmentation_3d.py [hunk 1]
```diff
@@ -112,9 +113,15 @@ def _loss_fn(i, j):
 
 @trainer.on(Events.EPOCH_COMPLETED)
 def log_training_loss(engine):
+    writer.add_scalar('Loss/train', engine.state.output, engine.state.epoch)
+    first_image_tensor = engine.state.batch[0][0]
```
Comment (Nic-Ma) [inline examples/unet_segmentation_3d.py]: I think you should show the **model output** instead of the original image, or both.
Just add an output_transform to SupervisedTrainer in this example, the default transform is:
```py
output_transform: Callable = lambda x, y, y_pred, loss: loss.item())
```
You can get all things: _image(x), label(y), output(y_pred)_

File: monai/utils/img2tensorboardutils.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+def _image3_animated_gif(imp, scale_factor=1):
+    """
+    Function to actually create the animated gif.
+    Args:
+        imp: tuple of tag and a list of image tensors
+        scale_factor: amount to multiply values by (if the image data is between 0 and 1, using 255 for this value will
+        scale it to displayable range)
+    """
+    from PIL.GifImagePlugin import Image as GIF
```
Comment (Nic-Ma) [inline monai/utils/img2tensorboardutils.py]: Suggest to move "import" at the beginning.

File: requirements.txt [hunk 3]
```diff
@@ -13,3 +13,4 @@ pandas
 coverage
 nibabel
 parameterized
+tensorboard
```
Comment (Nic-Ma) [inline requirements.txt]: I think we don't need to install tensorboard anymore as PyTorch already supported it, right?

Comment (nvkevlu) [reply requirements.txt]: this is still needed in order to actually use the functionality

File: monai/utils/img2tensorboardutils.py (hunk 2 above)
Comment (nvkevlu) [reply monai/utils/img2tensorboardutils.py]: Done

**Comment by Nic-Ma** [comment]
@nvkevlu ，

BTW, about the flake8 error, instead of **GIF**, I think maybe you can try:
```py
from XXX import XXX as Gif
```
or
```py
from XXX import XXX as gif
```
Please feel free to ask Eric, Ben, Wenqi or me for review again when you resolved the comments.
Thanks.

**Comment by nvkevlu** [comment]
> @nvkevlu ，
> 
> BTW, about the flake8 error, instead of **GIF**, I think maybe you can try:
> 
> ```python
> from XXX import XXX as Gif
> ```
> 
> or
> 
> ```python
> from XXX import XXX as gif
> ```
> 
> Please feel free to ask Eric, Ben, Wenqi or me for review again when you resolved the comments.
> Thanks.

The flake8 error is still not resolved after trying the suggested changes. After some searching, it seems this rule may be one of the ones that should be ignored.

As for model outputs, I have now modified the code to log 3 model outputs as well as the labels to tensorboard. I can imagine further work done to allow for contours to be overlayed over the image to make a more useful graphic. Also, the util functions can still be improved to have a parameter handle batch dimension as well as other shape formats of data. For some reason the results are still not exactly what I am expecting and I am not entirely sure why:

- It seems that the cropped regions look the same for each epoch even though UniformRandomPatch is used as a transform, why? Is this expected?
- Even after trying to use torch.where to reduce the labels to a single class, for some reason the label gifs can appear completely blank. The behavior seems inconsistent to me and I am stumped trying to figure out why this is happening.

**Comment by ericspod** [comment]
Is it possible to optionally use tensorboardx in place of Pytorch's implementation? @atbenmurray has encountered issues with Pytorch's SummaryWriter.

**Comment by wyli** [comment]
> * Even after trying to use torch.where to reduce the labels to a single class, for some reason the label gifs can appear completely blank. The behavior seems inconsistent to me and I am stumped trying to figure out why this is happening.

perhaps you need to tune the scaling factor of `_image3_animated_gif`:
https://github.com/Project-MONAI/MONAI/blob/fa62d5e4d6fb5d5672aa03f3b61765fca107bef4/monai/utils/img2tensorboardutils.py#L12-L13

**Comment by nvkevlu** [comment]
> Is it possible to optionally use tensorboardx in place of Pytorch's implementation? @atbenmurray has encountered issues with Pytorch's SummaryWriter.

Not sure if there is an easy and clean way to do that without changing the imports. Do you know what the issues were? From what I read, tensorboardx is where Pytorch's torch.utils.tensorboard came from, and it seems older versions of Pytorch had problems that were resolved in later versions.

**Comment by nvkevlu** [comment]
> > * Even after trying to use torch.where to reduce the labels to a single class, for some reason the label gifs can appear completely blank. The behavior seems inconsistent to me and I am stumped trying to figure out why this is happening.
> 
> perhaps you need to tune the scaling factor of `_image3_animated_gif`:
> 
> https://github.com/Project-MONAI/MONAI/blob/fa62d5e4d6fb5d5672aa03f3b61765fca107bef4/monai/utils/img2tensorboardutils.py#L12-L13

I have been using 255 for the scaling factor, which should scale 0 and 1 to 0 and 255. The where before that should be making all the values 0 and 1.

File: examples/unet_segmentation_3d.py (hunk 1 above)
Comment (nvkevlu) [reply examples/unet_segmentation_3d.py]: The model output is now being logged in the latest version.

**Comment by Nic-Ma** [comment]
> > Is it possible to optionally use tensorboardx in place of Pytorch's implementation? @atbenmurray has encountered issues with Pytorch's SummaryWriter.
> 
> Not sure if there is an easy and clean way to do that without changing the imports. Do you know what the issues were? From what I read, tensorboardx is where Pytorch's torch.utils.tensorboard came from, and it seems older versions of Pytorch had problems that were resolved in later versions.

Hi @ericspod ,

Could you please help show more details of the issue Ben faced before?
And as this task is for TensorBoard, maybe we can open another task for **TensorBoardX** in a later sprint if necessary. What do you think?

Hi @nvkevlu ,

(1) I think you can just modify the **.flake8** config file to ignore this format check issue.
(2) And paste some **screenshots** of your TensorBoard in a comment or PR description would be wonderful.
(3) Could you please help print some logs about the random crop issue you faced?
     Maybe @ericspod  can help to resolve it.
Thanks.

**Comment by Nic-Ma** [comment]
@nvkevlu ,

I already fixed your previous flake8 issues, please take a look and focus on other pending questions.
Thanks.

**Comment by nvkevlu** [comment]
Screenshot added showing tensorboard with logged gif of model output and label.

I don't actually have any more code changes at this time. Results are looking pretty much in line with what I expect now.

File: examples/unet_segmentation_3d.py [hunk 4]
```diff
@@ -104,17 +105,40 @@ def _loss_fn(i, j):
...
 trainer.add_event_handler(event_name=Events.EPOCH_COMPLETED, handler=checkpoint_handler, to_save={'net': net})
 
 
 @trainer.on(Events.EPOCH_COMPLETED)
 def log_training_loss(engine):
-    print("Epoch", engine.state.epoch, "Loss:", engine.state.output)
+    writer.add_scalar('Loss/train', engine.state.output[1], engine.state.epoch)
+
+    ones = torch.ones(engine.state.batch[1][0].shape, dtype=torch.int32)
+    first_image_tensor = engine.state.output[0][1][0].detach().cpu()
```
Comment (Nic-Ma) [inline examples/unet_segmentation_3d.py]: Seems it's **output** data instead of **image**, I think "image" is confusing for input.
Can we change it to "first/second/third_output_tensor"?
Thanks.

File: monai/utils/img2tensorboardutils.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        scale_factor: amount to multiply values by (if the image data is between 0 and 1, using 255 for this value will
+        scale it to displayable range)
+        global_step: Global step value to record
+    """
+    writer._get_file_writer().add_summary(make_animated_gif_summary(tag, image_tensor, max_out=max_out,
+                                                                    animation_axes=[1], image_axes=[2, 3],
+                                                                    scale_factor=scale_factor), global_step)
+
+
+def add_animated_gif_no_channels(writer, tag, image_tensor, max_out, scale_factor, global_step=None):
```
Comment (Nic-Ma) [inline monai/utils/img2tensorboardutils.py]: If this GIF feature only works for CDHW data, why not **combine** add_animated_gif() and add_animated_gif_no_channels()?

I think:
option 1: you can get the shape of input data and add 1 channel dim if not existing,
option 2: force the input to be CDHW and ask user to convert their data to be CDHW.
What do you think?

Comment (nvkevlu) [reply monai/utils/img2tensorboardutils.py]: I haven't actually made the decision to only support CDHW. In fact, I would like to incorporate batch with a parameter to specify how many items of the batch to load in a later version.

Comment (Nic-Ma) [reply monai/utils/img2tensorboardutils.py]: Hi @nvkevlu ,

OK, thanks for your explanation.
But why you need to show 2D image in **GIF** format? So I thought this tool only works for 3D/CDHW.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod and @wyli ,

I think this PR almost looks good to me.
Could you please also help review it?
Thanks.

File: monai/utils/img2tensorboardutils.py (hunk 5 above)
Comment (nvkevlu) [reply monai/utils/img2tensorboardutils.py]: 2D won't be supported but CDHW and NDHW may be.

Comment (Nic-Ma) [reply monai/utils/img2tensorboardutils.py]: Hi @nvkevlu ,

At least for this MVP version MONAI, we finalized to only support CDHW, no need to consider other shape formats.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @ericspod ,

Could you please also take a look at this PR?
Thanks.
